### PR TITLE
Feature/result cast

### PR DESCRIPTION
**Pull request by Jammy2211**
New API for how the `Analysis` class returns a `Return`.

Now, the `Analysis` class has a `Result` class attribute, which a user can overwrite with their own custom `Result` class. 

This removes a number of unecessary `make_result` functions that end up populating `Analysis` objects and is a cleaner API for users, as shown in the docs updated in this PR.

### CODE REVIEW

File: autofit/non_linear/result.py [hunk 1]
```diff
@@ -163,50 +198,141 @@ def model_bounded(self, b: float) -> AbstractPriorModel:
...
+        if self._samples is not None:
+            return self._samples
+
+        try:
+            Samples.from_csv(
+                paths=self.paths,
+                model=self.model,
+            )
+        except FileNotFoundError:
+            pass
```
Comment (rhayes777) [inline autofit/non_linear/result.py:311]: Did you mean to return this?

File: test_autofit/interpolator/test_covariance.py [hunk 2]
```diff
@@ -31,89 +31,89 @@ def test_covariance_matrix(interpolator):
...
+#         interpolator.covariance_matrix(), interpolator.inverse_covariance_matrix()
+#     )
+#     print(identity)
+#     assert np.allclose(
+#         identity,
+#         np.eye(9),
+#     )
+
+
+# def test_interpolate(interpolator):
```
Comment (rhayes777) [inline test_autofit/interpolator/test_covariance.py:45]: Ooof why is this all commented out?

**Review by rhayes777 — APPROVED** [review]
We need to fix those interpolator tests

File: autofit/example/analysis.py [hunk 3]
```diff
@@ -43,6 +53,34 @@ def log_likelihood_function(self, instance: af.ModelInstance) -> float:
         -------
         The log likelihood value indicating how well this model fit the dataset.
         """
+        model_data_1d = self.model_data_1d_from(instance=instance)
+
+        residual_map = self.data - model_data_1d
+        chi_squared_map = (residual_map / self.noise_map) ** 2.0
+        log_likelihood = -0.5 * sum(chi_squared_map)
```
[the remark this answers is not in the document]
Comment (Jammy2211) [reply autofit/example/analysis.py:60]: This is for the example module, which users might read, so intentionally simple.

**Pull request merged**